[Chen Yu An] iP

putaojuice opened the pull request:
# DukeTaskManager

> "_Every time your mind shifts, your world shifts._" - Byron Katie ([source](https://kidadl.com/articles/best-clear-mind-quotes-to-let-you-collect-your-thoughts))

**DukeTaskManager** frees your mind of having to remember things you need to do. It's,
- text-based
- easy to pick up
- **SUPER FAST** to use

All you need to do is,
1. download it ([here](https://github.com/putaojuice/ip))
2. double-click it.
3. add your tasks.
4. let it manage your tasks and free your mind 😄 

And it is **FREE**!

Current features:
- [x] Managing tasks
- [ ]  Managing deadlines (coming soon)
- [ ]  Reminders (coming soon)

If you are Java programmer, you can use it to practice Java too. Here's the `main` method:
```java
public class Duke {
    public static void main(String[] args)  {
        new Duke().startProgram();
    }
}
```

Commit:
Add Gradle support

putaojuice committed:
Duke Level-1

putaojuice committed:
Duke Level-2

putaojuice committed:
Duke Level-3

putaojuice committed:
Duke Level-4. Restructured the code to make everything cleaner and closer to OOP

putaojuice committed:
A-TextUiTesting

putaojuice committed:
Duke Level-5

putaojuice committed:
Duke Level-6
- Add Delete operation
- Add exception handling for IndexOutOfBound
- Handle negative number for mark/unmark/delete operations
- Abstract user input into a separate class

putaojuice committed:
Duke Level-7

- Add Save feature.

putaojuice committed:
Revert "Duke Level-7"

This reverts commit e7b249a76583abd19251c65fad6df98aced4be73.

putaojuice committed:
Duke Level-7

putaojuice committed:
Duke Level-8

putaojuice committed:
Merge branch 'branch-Level-7'

putaojuice committed:
Merge branch 'branch-Level-8'

putaojuice committed:
Duke A-MoreOOP

Duke class: extract into multiple classes to make it more OOP

Duke class currently has a lot logic operations that should be abstract into other classes to maintain OOP paradigm.

Abstract all unnecessary methods that should not belong to Duke class into other classes: Ui, TaskList, Parser and Storage.

This is to maintain Duke class's singularity behaviour.

putaojuice committed:
Duke A-Packages

putaojuice committed:
Duke A-JUnit

- Add JUnit test for Duke
- Add JUnit test for TaskList
- Add JUnit test for Ui

putaojuice committed:
Package Fix

Fix path to exception class

putaojuice committed:
Exception Package Update

-Add all exception classes as same package in duke

putaojuice committed:
Update class import statements

putaojuice committed:
Duke A-JavaDoc

- Add JavaDoc to classes and methods

putaojuice committed:
Duke A-CodingStandard

- Edit code to stick to Java coding standard

putaojuice committed:
Merge branch 'branch-A-CodingStandard'

putaojuice committed:
Duke Level-9

- Add method to find task with keyword

putaojuice committed:
Merge branch 'branch-Level-9'

putaojuice committed:
Merge branch 'add-gradle-support'

putaojuice committed:
A-Gradle

- Add Gradle to build and run Duke

putaojuice committed:
Duke A-CheckStyle

- Add checkStyle with Gradle to check for coding style violation

putaojuice committed:
Change import statement order

putaojuice committed:
Duke Level-10

- Add basic GUI to Duke

limweiliang commented on src/main/java/duke/Duke.java:
```diff
@@ -0,0 +1,330 @@
+/**
+ * Duke is a task tracker interactive chatbot.
+ *
+ * @author Chen Yu An
```
I liked how you included the author and version number here!

putaojuice replied:
Thank you 😄

limweiliang commented on src/main/java/duke/Duke.java:
```diff
@@ -0,0 +1,330 @@
+ */
+public class Duke extends Application {
+    private final Scanner scanner = new Scanner(System.in);
+    private boolean endProgram = false; // state to terminate the program
```
Perhaps this could be named something like `isRunning` to better indicate that this is a boolean to better fit the coding standards?

putaojuice replied:
Yeap, you're right, will change it asap.

limweiliang commented on src/main/java/duke/Duke.java:
```diff
@@ -0,0 +1,330 @@
+        userInput.clear();
+    }
+
+    /**
```
Was this meant to be changed?

putaojuice replied:
Not really sure if we are not allowed to change it. I was trying to test out if there's anyway to exit the program after saying "bye". Will probably make it into another method.

limweiliang commented on src/main/java/duke/Duke.java:
```diff
@@ -0,0 +1,330 @@
+package duke;
```
This file is really long! Might I suggest breaking it down into more classes?

putaojuice replied:
Yes, I'll clean this class up after the GUI is properly implemented. Right now I know there are a lot unnecessary lines in it.

limweiliang commented on src/main/java/duke/io/Parser.java:
```diff
@@ -0,0 +1,111 @@
+     * @return Return a UserInput for Duke to process subsequent actions.
+     * @throws DukeWrongTimeFormatException If the user's input has invalid time format for certain type of task.
+     */
+    public UserInput parseInput(String input) throws DukeWrongTimeFormatException {
```
I liked how you divided up the method into clear, commented segments, but maybe this could be even clearer if it were divided into multiple methods?

putaojuice replied:
Yeap definitely, will think of a way to abstract it!

limweiliang commented on src/test/java/duke/task/TaskListTest.java:
```diff
@@ -0,0 +1,216 @@
+            assertEquals("________________________________________\n"
+                            + "Got it. I've added this task:\n" + "D][ ] proj (by: 2022-01-31)\n"
+                            + "Now you have 1 tasks in the list\n"
+                            + "________________________________________\n",
```
Maybe this could become a static constant?

putaojuice replied:
Yeah, sounds good.

limweiliang reviewed (commented):
LGTM. I think the code was generally okay, but some functions tended to be a bit long and complex, so I think it would help with readability, maintainability and testability if they were broken down into smaller functions.

Good job on following the code quality guidelines!

limweiliang commented on src/main/java/duke/task/DeadlineTask.java:
```diff
@@ -0,0 +1,29 @@
+ * Has extra time field for the deadline of this task.
+ */
+public class DeadlineTask extends Task {
+    private String time;
```
Maybe naming this `deadline` would be more informative?

putaojuice replied:
Right, makes more sense. Thanks for pointing that out!

AryanSarswat commented on src/main/java/duke/Duke.java:
```diff
@@ -0,0 +1,330 @@
+        AnchorPane.setBottomAnchor(sendButton, 1.0);
+        AnchorPane.setRightAnchor(sendButton, 1.0);
+
+        AnchorPane.setLeftAnchor(userInput , 1.0);
```
You could maybe remove the whitespace after userInput

AryanSarswat commented on src/main/java/duke/exception/DukeException.java:
```diff
@@ -0,0 +1,10 @@
+package duke.exception;
+
+/**
+ * An exception class specifically for Duke
```
You should consider adding a period at the end of the JavaDoc comment.

AryanSarswat commented on src/main/java/duke/task/Task.java:
```diff
@@ -0,0 +1,47 @@
+    }
+
+    /**
+     * Mark the task as done
```
Consider adding a period at the end of the javadoc comment

AryanSarswat commented on src/main/java/duke/Duke.java:
```diff
@@ -0,0 +1,330 @@
+        AnchorPane.setBottomAnchor(sendButton, 1.0);
+        AnchorPane.setRightAnchor(sendButton, 1.0);
+
+        AnchorPane.setLeftAnchor(userInput , 1.0);
```
Maybe remove whitespace after userInput?

putaojuice replied:
I am sorry I don't quite get what you mean to remove the whitespace after userInput. This line of code doesn't set any whitespace?

AryanSarswat replied:
so in line 207:
```Java
AnchorPane.setLeftAnchor(userInput , 1.0);
```

it should not have a whitespace after userInput like this:

```Java
AnchorPane.setLeftAnchor(userInput{removed whitespace here}, 1.0);
```

AryanSarswat commented on src/main/java/duke/task/Task.java:
```diff
@@ -0,0 +1,47 @@
+    }
+
+    /**
+     * Mark the task as done
```
Consider adding period at the end of the description. This applies to line 37 in this doc aswell.

putaojuice replied:
Noted. Thanks for pointing it out!

AryanSarswat commented on src/main/java/duke/task/ToDoTask.java:
```diff
@@ -0,0 +1,19 @@
+    public ToDoTask(String taskName) {
+        super(taskName);
+    }
+    public ToDoTask(String taskName, boolean isDone) {
```
Consider adding a line separation between the two constructors.

putaojuice replied:
Noted. Thanks for pointing it out!

AryanSarswat commented on src/test/java/duke/task/TaskListTest.java:
```diff
@@ -0,0 +1,216 @@
+    }
+
+    @Test
+    public void addTask_addDeadlineProject_exceptionThrownNoDescription() throws DukeException, IOException {
+        setUpStreams();
+        TaskList taskList = new TaskList();
+        UserInput testInput = new UserInput("D", "deadline", "","" , false);
+        try {
+            taskList.addTask(testInput);
+            assertEquals("________________________________________\n"
+                            + "Got it. I've added this task:\n" + "D][ ] proj (by: 2022-01-31)\n"
+                            + "Now you have 1 tasks in the list\n"
+                            + "________________________________________\n",
+                    outContent.toString());
+            fail(); // the test should not reach this line
+        } catch (DukeException e) {
+            assertEquals("OOPS!!! The description of a deadline cannot be empty.", e.getMessage());
+        }
+        restoreStreams();
+    }
+
+    @Test
+    public void addTask_addDeadlineProject_exceptionThrownNoTime() throws DukeException, IOException {
+        setUpStreams();
+        TaskList taskList = new TaskList();
+        UserInput testInput = new UserInput("D", "deadline", "proj ","" , false);
+        try {
+            taskList.addTask(testInput);
+            assertEquals("________________________________________\n"
+                            + "Got it. I've added this task:\n" + "D][ ] proj (by: 2022-01-31)\n"
+                            + "Now you have 1 tasks in the list\n"
+                            + "________________________________________\n",
+                    outContent.toString());
+            fail(); // the test should not reach this line
+        } catch (DukeException e) {
+            assertEquals("OOPS!!! Remember to set a time.", e.getMessage());
+        }
+        restoreStreams();
+    }
+}
```
Great job on the thorough testing! 😄

putaojuice replied:
Thank you!

AryanSarswat commented on src/main/java/duke/io/UserInput.java:
```diff
@@ -0,0 +1,90 @@
+
+    /**
+     * A method to get the time of a task.
+     * @return String of the time.
```
Consider adding a line separating your description and the ``@return param`` so it would be something like :
```Java
/**
     * A method to get the time of a task.
     * (Space to Consider adding)
     * @param blah
     * @return blah
**/ 
```

This is a general comment which applies to your other javadocs comments aswell!

putaojuice replied:
Noted. Thanks for pointing it out!

AryanSarswat commented on src/main/java/duke/io/UserInput.java:
```diff
@@ -0,0 +1,90 @@
+ * This class helps Duke to make sense of the user's input.
+ */
+public class UserInput {
+    private String type = "";
+    private String command = "";
+    private String description = "";
+    private String time = "";
```
Is there a reason you initialized these variables to empty strings? You could leave them uninitialized maybe?

putaojuice replied:
It's just my habit to make sure none of the variables are uninitialised in the later part of my code.

AryanSarswat commented on src/main/java/duke/task/TaskList.java:
```diff
@@ -0,0 +1,405 @@
+        if (userInput.getCommand().equals("deadline")) {
+            if (userInput.getDescription().equals("")) {
+                // throw no description exception
+                throw new DukeNoDescriptionException("OOPS!!! The description of a deadline cannot be empty.");
```
You could consider handling these exception in the constructor of the relevent tasks. i.e throws these exception while creating the task during ->
 ```Java 
new Deadline("{empty description}", "{incorrect time format}");
```

putaojuice replied:
Alright, thanks for the recommendation, will think about it.

putaojuice committed:
Update code according to PR comments

- Update the code according to peer feedbacks.

putaojuice committed:
Remove unreachable exceptions

There are some methods in the Storage class has exceptions that are
unreachable.

Remove all those exceptions that are being catch earler.

They are simply not needed in the program anymore.

Removing them will reduce confusion when someone else is looking at
the code.

putaojuice committed:
Duke A-Assertions

There is no assert statement used in the program to assert correctness.

Include Java assert in the program at critical point where certain data should
be enforce with certain properties for the correct behaviour.

Assert is applied mainly in the TaskList, Parser and Storage classes.

putaojuice committed:
Duke A-CodeQuality

The current code does not comply to Code Quality guideline.

Refactor multiple methods that does not comply to Code Quality guideline.
This includes reduce redundant if-else statements. Apply KISS principle to
certain methods that is currently doing too many things at once. Reduce long
methods. Try to follow SLAP principle.

It is done this way to keep methods short and easy to read. Reduce redundant
if-else statements helps to keep the code tidy and remove nested conditions.

Let's,
- Remove redundant methods in Ui, Duke, TaskList, Task classes
- Refactor methods in Parser, Storage, TaskList classes
- Add date format limitation to parseEventInput()
- Add bug fix to markDone(), markUndone() and deleteTask() where the user
could enter String instead of int.

putaojuice committed:
Merge pull request #1 from putaojuice/branch-A-Assertions

Duke A-Assertions

putaojuice committed:
Merge branch 'master' into branch-A-CodeQuality

putaojuice committed:
Merge pull request #2 from putaojuice/branch-A-CodeQuality

Duke A-CodeQuality

putaojuice committed:
Java documentation and coding standard

- Add various java documentation for constructors
- Edit code to adhere to java coding standard

putaojuice committed:
Duke BCD-Extension

-Implement extension C-Help

putaojuice committed:
Duke A-UserGuide

-Add a brief User Guide

putaojuice committed:
Minor update on README

-Update docs/README.md

putaojuice committed:
Fix checkstyle
